### ISSUE TITLE
Jump-over/moveright is broken for triple-quoted strings in Python

### ISSUE THREAD

**Issue by Diomendius** [windwp/nvim-autopairs#229]
### Description

When `filetype=python` and in insert mode with the cursor within a `"""`-delimited string before the closing quotes (at the `|` in `"""some text|"""`), pressing `"` causes the cursor to jump over only one of the closing `"` characters. Pressing `"` additional times causes one `"` character to be inserted on the first press, then for every subsequent press four `"`s are inserted (one to the left of the new cursor position and three to the right).

It looks like the `"""` rule doesn't actually have a `move_cond`, and adding a basic one seems to work, though it probably misses some edge cases:
```lua
require('nvim-autopairs').get_rule('"""'):with_move(function() return true end)
```

Incidentally, there is no `'''` pair included by default, though both `"""` and `'''` are valid Python syntax.

### Steps to reproduce

1. In an empty buffer, set `ft=python` and enter insert mode.
2. Type `"""` and observe that the closing `"""` is auto-inserted.
3. Type `"` and observe only one closing `"` is jumped over.
4. Optionally, continue typing `"` and receive as many `"""` as you want without escaping the `"""string"""`.

### Minimal config

```Lua
vim.cmd [[set runtimepath=$VIMRUNTIME]]
vim.cmd [[set packpath=/tmp/nvim/site]]
local package_root = '/tmp/nvim/site/pack'
local install_path = package_root .. '/packer/start/packer.nvim'
local function load_plugins()
  require('packer').startup {
    {
      'wbthomason/packer.nvim',
      {
        'windwp/nvim-autopairs',
      },
      -- ADD PLUGINS THAT ARE _NECESSARY_ FOR REPRODUCING THE ISSUE
    },
    config = {
      package_root = package_root,
      compile_path = install_path .. '/plugin/packer_compiled.lua',
      display = { non_interactive = true },
    },
  }
end
_G.load_config = function()
  require('nvim-autopairs').setup()
end
if vim.fn.isdirectory(install_path) == 0 then
  print("Installing nvim-autopairs and dependencies.")
  vim.fn.system { 'git', 'clone', '--depth=1', 'https://github.com/wbthomason/packer.nvim', install_path }
end
load_plugins()
require('packer').sync()
vim.cmd [[autocmd User PackerComplete ++once echo "Ready!" | lua load_config()]]
```

**Comment by Diomendius** [comment]
I think I understand the issue with jump-over being triggered by unrelated bindings now. It is definitely an issue I encountered before, but I didn't look deeply enough into its cause at the time. A rule's `with_move` / `move_cond` callbacks actually get called whenever any mapped key is pressed when the cursor is before the closing pair for the rule, regardless of whether the actual key pressed matches the rule:
https://github.com/windwp/nvim-autopairs/blob/06535b1f1aefc98df464d180efa693bb696736c4/lua/nvim-autopairs.lua#L448-L451
This means that if you set a move condition for a rule, you currently need to check that `opts.char` matches the key you want to trigger the move. This seems unintuitive and doesn't seem to be documented, but it might not be an oversight as `cond.move_right` already does this check:
https://github.com/windwp/nvim-autopairs/blob/06535b1f1aefc98df464d180efa693bb696736c4/lua/nvim-autopairs/conds.lua#L227-L230
This doesn't work for multi-character closing sequences, by the way, but this can be fixed by changing `opts.next_char` to `opts.next_char:sub(1, 1)`, at least so long as the first character of `next_char` is ASCII, otherwise Lua will not make your life easy. Actually, `opts.next_char:sub(1, #opts.char)` would probably work, but if you need Unicode support for auto pairs you're probably using it for the wrong thing. Or maybe you want to have auto pairs for `«»`, I don't know.

Anyway, it seems to work fine for now with this in my config:
```lua
require('nvim-autopairs').get_rule('"""'):with_move(function(opts)
    return opts.char == opts.next_char:sub(1, 1)
end)
```
Ideally it should be smarter than this. `"""`-delimited strings can have unescaped `"` within them, except at the beginning or end (`"""a string with "quotes" in it"""` parses fine, but `"""quote at the end""""` doesn't), though I'm not sure this causes problems on its own, as singular `"` characters don't currently expand when typing between `"""`. On the other hand, the fact that you can have unescaped `"`s within `"""`-strings means it could be undesirable to have a single `"` cause jump-over at all, for the same reasons as #228.

The next best thing might be to prevent a single `"` from causing jump-over within a `"""`-string and have `"""` automatically clear itself and jump over the closing `"""` if typed before the closing `"""`, but I'm not sure if this can be done cleanly with the current rules interface, and even if it could it might be too complex to have as default behaviour. `"""` can also be escaped (`\"""`), and as in #228 it should be possible to type this without triggering jump-over.

**Comment by windwp** [comment]
maybe the variable name `next_char` is make you think it is only one character and the key press but it is not.
It combine the key press and the string after the cursor so if you press `"` then next_char is `"""`.

https://github.com/windwp/nvim-autopairs/blob/06535b1f1aefc98df464d180efa693bb696736c4/lua/nvim-autopairs.lua#L448-L451
it need to compare the end pair with next_char to check the rule because the default result of function can_move() is true (if it doesn't have any condition return false)

nvim-autopairs support autopairs with unicode character i don't test for multiple of unicode character but it have a test for 1 unicode character pairs.

https://github.com/windwp/nvim-autopairs/blob/master/tests/nvim-autopairs_spec.lua#L491-L505

btw: you can get the length of unicode string by vim.api.nvim_strwidth()

**Comment by Diomendius** [comment]
> maybe the variable name `next_char` is make you think it is only one character and the key press but it is not. It combine the key press and the string after the cursor so if you press `"` then next_char is `"""`.

Yes, the name is not ideal when the closing pair is multiple characters, but I did realize this. My point is that `char` is not checked at all by `autopairs_map`, so if a `move_cond` exists and returns `true` without checking that `char` matches what the rule expects, *any* key mapped by `nvim-autopairs` will cause jump-over so long as the characters after the cursor match `rule.end_pair`. i.e. if the `"{", "}"` rule was defined with only one move condition via `:with_move(cond.done())`, pressing `]`, `)`, `"` (and even openers such as `{`, `(`, `[`) would cause jump-over/moveright when the character after the cursor is `}`.

> btw: you can get the length of unicode string by vim.api.nvim_strwidth()

This is useful but not for this problem. `nvim_strwidth(str)` returns how many display cells the characters of `str` occupy, but what I needed to do was test if `opts.char` matches the start of `opts.next_char`, hence `opts.char == opts.next_char:sub(1, #opts.char)`. The relevant function would be one that returns how many UTF-8 bytes a string occupies, but that's just the length of the string, which I didn't think of at first. A function that returned the first (or nth) unicode character of a string would also be useful, but since `char` should match the keybinding (I think even if the binding is multiple key presses) it's much easier to do it this way.
